python/src/udf.rs

```diff
@@ -1,32 +1,131 @@
-use arrow::array::ArrayData;
+use std::fmt::Debug;
+use std::sync::Arc;
+use itertools::Itertools;
+
+use arrow::array::{ArrayData, ArrayRef};
 use arrow::datatypes::DataType;
 use arrow::pyarrow::{FromPyArrow, ToPyArrow};
 use pyo3::prelude::*;
 use pyo3::types::PyTuple;
+use sparrow_instructions::{Evaluator, RuntimeInfo, StaticInfo, ValueRef};
+use sparrow_syntax::{FeatureSetPart, Signature};
+use uuid::Uuid;
+
+#[derive(Debug)]
+struct PyUdf {
+    signature: Signature,
+    callable: Py<PyAny>,
+    uuid: uuid::Uuid,
+}
+
+struct PyUdfEvaluator {
+    args: Vec<ValueRef>,
+    callable: Py<PyAny>,
+}
+
+#[derive(Clone)]
+#[pyclass]
+// TODO: Does it need to extend Expr for some reason? 
```

Comment:
I ask this because I don't understand why `Table` is a subclass of `Expr`

Reply:
I think table did because we wanted to allow every source to be used as an expression. We don't *need* to model that in the Rust code -- the `source` could hold both... but at the least, we needed the sources to extend `Timestream`. I'm not sure if it would be clearer to have them separate in the Rust code -- we could say:

* `Timestream has the _ffi.Expr (and/or rename that to Timestream)`
* `Source extends Timestream (so it has an _ffi.Expr) and it has a _ffi.Table (which we should rename to Source)`.

The nice thing if we split them like that is that the Rust code doesn't deal with subtyping, but also it keeps the separation between expressions and sources more distinct in rust.

Reply:
Okay, I think that makes sense. `Table` extends `Expr`, which is the object within `Timestream` currently. Yeah, if we need that separation, that would be a good idea.